### ISSUE TITLE
SQLite backend can flood database

### ISSUE THREAD

**Issue by layday** [requests-cache/aiohttp-client-cache#136]
### The problem
The SQLite backend establishes a new connection to the database with every request.  This means that, given a sufficiently large number of concurrent connections, you might end up hitting the file descriptor limit of your OS, which for my OS (macOS Monterey) is 256.  Subsequent connections to the database will fail with "unable to open database file".

### Expected behavior
Preferably, there should be some way to persist the SQLite connection to the database, while still committing changes in real time (i.e. not `bulk_commit`).  If nothing else, opening a new connection per request seems a little wasteful and I assume is a couple of orders of magnitude slower than querying an already open database would be.

### Steps to reproduce the behavior
```py
import asyncio
import sqlite3

from aiohttp_client_cache import CachedSession, SQLiteBackend


async def make_request(session: CachedSession, attempt: int):
    try:
        async with session.get("http://example.com/"):
            pass
    except sqlite3.OperationalError:
        print(f"errored after {attempt} attempts")
        raise


async def main():
    async with CachedSession(cache=SQLiteBackend("foo.db")) as session:
        await asyncio.gather(*(make_request(session, a) for a in range(256)))


asyncio.run(main())

```

Running this produces:

```sh
errored after 255 attempts
[...]
Traceback (most recent call last):
  File "test.py", line 21, in <module>
    asyncio.run(main())
  File "/nix/store/pgnhckqsc2n2aspvq2pvfvpkw3bgm2nd-python3-3.11.0/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/nix/store/pgnhckqsc2n2aspvq2pvfvpkw3bgm2nd-python3-3.11.0/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/pgnhckqsc2n2aspvq2pvfvpkw3bgm2nd-python3-3.11.0/lib/python3.11/asyncio/base_events.py", line 650, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "test.py", line 18, in main
    await asyncio.gather(*(make_request(session, a) for a in range(256)))
  File "test.py", line 9, in make_request
    async with session.get("http://example.com/"):
  File "/[...]/aiohttp/client.py", line 1141, in __aenter__
    self._resp = await self._coro
                 ^^^^^^^^^^^^^^^^
  File "/[...]/forge/_revision.py", line 322, in inner
    return await callable(*mapped.args, **mapped.kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/[...]/aiohttp_client_cache/session.py", line 46, in _request
    response, actions = await self.cache.request(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/[...]/aiohttp_client_cache/backends/base.py", line 127, in request
    response = None if actions.skip_read else await self.get_response(actions.key)
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/[...]/aiohttp_client_cache/backends/base.py", line 135, in get_response
    response = await self.responses.read(key) or await self._get_redirect_response(str(key))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/[...]/aiohttp_client_cache/backends/sqlite.py", line 182, in read
    return self.deserialize(await super().read(key))
                            ^^^^^^^^^^^^^^^^^^^^^^^
  File "/[...]/aiohttp_client_cache/backends/sqlite.py", line 153, in read
    async with self.get_connection() as db:
  File "/nix/store/pgnhckqsc2n2aspvq2pvfvpkw3bgm2nd-python3-3.11.0/lib/python3.11/contextlib.py", line 204, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/[...]/aiohttp_client_cache/backends/sqlite.py", line 73, in get_connection
    else await aiosqlite.connect(self.filename, **self.connection_kwargs)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/[...]/aiosqlite/core.py", line 137, in _connect
    self._connection = await future
                       ^^^^^^^^^^^^
  File "/[...]/aiosqlite/core.py", line 102, in run
    result = function()
             ^^^^^^^^^^
  File "/[...]/aiosqlite/core.py", line 397, in connector
    return sqlite3.connect(loc, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlite3.OperationalError: unable to open database file
```

### Workarounds
Using a synchronisation primitive to limit concurrent connections.

### Environment
- aiohttp-client-cache version: 0.7.3
- Python version: 3.11
- Platform: macOS

---

Thanks for `aiohttp-client-cache`!

**Comment by AIGeneratedUsername** [comment]
I tried to run your code snippet with `2_000` instead of `256`, and I was not able to get an error.

I did a quick look at the code here https://github.com/requests-cache/aiohttp-client-cache/blob/73fbd3988b13c8f1193019c29ed7a0523004bbde/aiohttp_client_cache/backends/sqlite.py#L73 and it looks like a connection should be opened once. Perhaps the package creator may clarify this better.

PS: sqlite is not designed for high concurrency, so (if possible) you may want to look on other backends for a high load or stress testing

**Comment by layday** [comment]
I was hitting this in a scraper, not anything that needs high concurrency :)

> I tried to run your code snippet with `2_000` instead of `256`, and I was not able to get an error.

Do you use macOS?

**Comment by AIGeneratedUsername** [comment]
Sorry, I overlooked that you pointed to a specific OS. No, I did not test with your OS. I tested only on Linux.

**Comment by JWCook** [comment]
Thanks for the details. Yes, I can change this to reuse the same connection object. This is how requests-cache behaves (using stdlib `sqlite3`), but it seems I didn't yet port that to this library.

I'll just need to do a bit more testing to see if `aiosqlite` has any different behavior regarding thread safety, leaving open connections without closing them, etc.

**Comment by JWCook** [comment]
> I assume is a couple of orders of magnitude slower than querying an already open database would be.

It is a bit slower, but not by orders of magnitude. IIRC it's somewhere in the range of tens to a couple hundred microseconds to open a SQLite connection. The majority of the time spent per write is in the `COMMIT` (1-2 milliseconds depending on disk speed).

The main part I'm surprised about here is that `connection.close()` doesn't release the file descriptor.

**Comment by layday** [comment]
I think it’s simply that not enough coros reach close before hitting the fd limit.

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------

**Comment by JWCook** [comment]
Ah, you're right.

Also, I was slightly off on connection time. It's actually quite a bit different between `sqlite3` and `aiosqlite`: 
```python
import asyncio
import aiosqlite
import sqlite3
from time import time

async def test_aiosqlite_connect():
    start = time()
    for i in range(255):
        db = await aiosqlite.connect('test.db')
        await db.close()
    elapsed = time() - start
    average = int((elapsed/255) * 1000000)
    print(f'Elapsed: {elapsed:.4f}s (average {average}µs per connection)')

def test_sqlite_connect():
     start = time()
     for i in range(255):
         db = sqlite3.connect('test.db')
         db.close()
     elapsed = time() - start
     average = int((elapsed/255) * 1000000)
     print(f'Elapsed: {elapsed:.4f}s (average {average}µs per connection)')

asyncio.run(test_aiosqlite_connect())
test_sqlite_connect()
```

On my machine, I get an average of 22µs for `sqlite3` and 450µs for `aiosqlite`.

**Comment by layday** [comment]
huh, I wonder why sqlite3 is so much faster for you:

```
Elapsed: 0.1149s (average 450µs per connection)
Elapsed: 0.0424s (average 166µs per connection)
```

**Comment by JWCook** [comment]
This has been fixed in 0.8.0. Thanks again for the bug report and suggestions!
